lib/archethic_cache/lru_disk.ex

```diff
@@ -0,0 +1,66 @@
+defmodule ArchethicCache.LRUDisk do
+  @moduledoc """
+  Wraps the LRU genserver and adds hooks to write / read from disk.
+  The value is always a binary.
+  """
+  alias ArchethicCache.LRU
+
+  require Logger
+
+  @spec start_link(GenServer.name(), non_neg_integer(), binary()) ::
+          {:ok, binary()} | {:error, term()}
+  def start_link(name, max_bytes, cache_dir) do
+    cache_dir = Path.join(cache_dir, "#{name}")
+    :ok = reset_directory(cache_dir)
+
+    LRU.start_link(name, max_bytes,
+      put_fn: fn key, value ->
+        # write to disk
+        File.write!(key_to_path(cache_dir, key), value, [:exclusive, :binary])
+
+        # store a nil value in the LRU's ETS table
+        nil
+      end,
+      get_fn: fn key, nil ->
+        # called only if the key is already in LRU's ETS table
+        case File.read(key_to_path(cache_dir, key)) do
+          {:ok, bin} ->
+            bin
+
+          {:error, _} ->
+            nil
+        end
+      end,
+      evict_fn: fn key, nil ->
+        case File.rm(key_to_path(cache_dir, key)) do
+          :ok ->
+            :ok
+
+          {:error, _} ->
+            :ok
+        end
+      end
+    )
+  end
+
+  @spec put(GenServer.server(), term(), binary()) :: :ok
+  defdelegate put(server, key, value), to: LRU, as: :put
+
+  @spec get(GenServer.server(), term()) :: nil | binary()
+  defdelegate get(server, key), to: LRU, as: :get
+
+  @spec purge(GenServer.server()) :: :ok
+  defdelegate purge(server), to: LRU, as: :purge
+
+  defp reset_directory(dir) do
+    File.rm_rf!(dir)
+    File.mkdir_p!(dir)
+  end
+
+  defp key_to_path(cache_dir, key) do
+    Path.join(
+      cache_dir,
+      Base.encode64(:erlang.term_to_binary(key))
```

Comment:
The file name is depending on the size of the requested url, so the file name can be very long. Maybe we should hash the key before encoding it in base 64, so the file name will have the same size no matter their url.
for example:
```elixir
addr = :crypto.strong_rand_bytes(34) |> Base.encode16
key = {addr, ["path", "to", "content"]}
Base.encode64(:erlang.term_to_binary(key))
# "g2gCbQAAAEQ4RkI4QkQ5RTkyRjVGRTZEODU0OTY5NDU1RTdGOEVBMDNCODQ0MEI2ODJEQkNFNUI2RkU0RjQ4MEFDRUQwMzc5NUNBQWwAAAADbQAAAARwYXRobQAAAAJ0b20AAAAHY29udGVudGo="

# Using a hash before
:crypto.hash(:sha256, :erlang.term_to_binary(key)) |> Base.encode64
# "mKj3eScg2Xh/949NXi59NjdHeNrhI4Zm3GuUv5Ehpis="
```

Reply:
Also we should use `Base.url_encode64` as the classic `Base.encode64` may add "/" in the file name which is a forbidden character

Reply:
What is the initial reason of the base64 ? why not hex ?

Reply:
> What is the initial reason of the base64 ? why not hex ?

no reason, 64 will return a shorter string than 16, should I switch to 16?